### ISSUE TITLE
Brancheable Storage

### ISSUE THREAD

**Issue by jhprinz** [choderalab/openpathsampling#282]
Funny that I always have ideas for storage. This one was inspired by talking to @fabian-paul:

When storing immutable objects, he said it would be good to go back and try different things. Not sure, if we really need this ourselves, but it could be interesting for using the storage for other things / simulations, too.

What he meant is, that you might want to simulate or create objects using a specific order parameter and later decide, well, no that was stupid and go back to the position in the storage before doing the useless simulations. Like in github making a branch of the simulation. First do the bootstrap, commit, equilibration, commit. Then branch into try1 and try2. Later analyze only branch1 or branch2 separately.
The way I would do this now is to have separate storages which works fine. 

The other way would be to add commits or tags at certain stages and since no elements get deleted, this is practically for free. All you need to store is the current number of objects of each type in the storage, which is the length of all tables. The positions between tags then mark your "commit". You could even later split the storage easily into separate files without worrying about missing items. Objects in different branches cannot rely on each other. 

That might be something for later as a gimmick. Just wanted to make sure I do not forget about it.

**Comment by dwhswenson** [comment]
Again, I think the more natural workflow here is just to use multiple files. But as a future-project thing, I have no problem supporting workflows that I won't personally use.

Doesn't this just amount to storing a dictionary of `{ tag_name : { table_name : length } }`? Low priority, but that doesn't sound too hard.

**Comment by jhprinz** [comment]
Exactly. Maybe this could be nice for archiving, etc... Very low priority but maybe nice to have and easy to implement.
